### PR TITLE
Fix #10764: delete url rewrites for products after an attribute set is deleted

### LINKED ISSUE #10764: Delete Attribute Set is not Deleting  product url key in  url rewrite
### Preconditions

1.  Magento 2.1.7
2. With needed prerequisites.

### Steps to reproduce

1.  Create a new attribute set ex: testattset
2.  Create a product for the newly created attribute set (test as product so URL key also become test)
3.  Delete the newly created attribute set, in M2 soon after attribute set is deleted all related product of that attribute set also deleted that is test product also deleted 
4.  But when you try to create a product with the same name it will throw error "URL key for specified store already exists." because URL key of that product is not deleted 

### Expected result

1.  when attribute set is deleted  all the product of that attribute set and related URL key from URL rewrite should also be deleted and allow to create same product the second time also 

### Actual result

1. its throw error "URL key for specified store already exists."

### PR DESCRIPTION
**Pull request by patrikpihlstrom**
### Description
Adds an interceptor around the delete function of `\Magento\Eav\Model\AttributeSetRepository` in order to capture the relevant product ids, and then remove the old url rewrites after the attribute set has been deleted.

### Fixed Issues (if relevant)

1. magento/magento2#10764: Delete Attribute Set is not Deleting product url key in url rewrite

### Manual testing scenarios
1. Create an attribute set
2. Create a product from the newly created attribute set and insure that a url key is specified
3. Delete the attribute set
4. Create a new product from the default attribute set and apply the same url key
5. Hopefully you won't get an error when you save the product

### CODE REVIEW

File: app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: patrikpihlstrom
+ * Date: 11/10/17
+ * Time: 22:12
+ */
+
+namespace Magento\CatalogUrlRewrite\Plugin\Eav\Model;
```
Comment (vrann) [inline app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php]: Please fix the namespace. Here is the test result:

```Magento\Test\Integrity\ClassesTest::testClassReferences
Passed: 24399, Failed: 1, Incomplete: 0, Skipped: 0.
Data set: /opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php
Incorrect namespace usage(s) found in file /opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php:
CouldNotDeleteException
/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/dev/tests/static/testsuite/Magento/Test/Integrity/ClassesTest.php:598
/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/dev/tests/static/testsuite/Magento/Test/Integrity/ClassesTest.php:371
/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/dev/tests/static/testsuite/Magento/Test/Integrity/ClassesTest.php:373

/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:115
/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:70
/opt/bamboo/bamboo-agent-home/xml-data/build-dir/MCCE22-AT398-INSE/magento2ce/dev/tests/static/testsuite/Magento/Test/Integrity/ClassesTest.php:373```

File: app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     * @param \Magento\Eav\Model\AttributeSetRepository $subject
+     * @param callable $proceed
+     * @param AttributeSetInterface $attributeSet
+     */
+    public function aroundDelete(
+        \Magento\Eav\Model\AttributeSetRepository $subject,
+        callable $proceed,
+        AttributeSetInterface $attributeSet
+    ) {
+        if (!isset($subject) || !$attributeSet->getAttributeSetId())
```
Comment (vrann) [inline app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php]: please fix the coding style

```
Magento\Test\Php\LiveCodeTest::testCodeStyle
PHP Code Sniffer detected 2 violation(s): 

FILE: ...to/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 50 | ERROR | [x] Expected 1 space after closing parenthesis; found 9
----------------------------------------------------------------------
```

**Review by vrann — DISMISSED** [review]

**Comment by vrann** [comment]
@patrikpihlstrom please make needed changes in order to accept the PR

File: app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php (hunk 1 above)
Comment (patrikpihlstrom) [reply app/code/Magento/CatalogUrlRewrite/Plugin/Eav/Model/AttributeSetRepository.php]: I'm not quite sure what the namespace is supposed to be if not `Magento\CatalogUrlRewrite\Plugin\Eav\Model`. I've copied the namespace structure from other plugins, for example `Magento\Catalog\Plugin\Model\ResourceModel`.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=11391) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=11391) before we can accept your contribution.<br/>**6** out of **13** committers have signed the CLA.<br/><br/>:white_check_mark: okorshenko<br/>:white_check_mark: vkublytskyi<br/>:white_check_mark: andrew-garside-temando<br/>:white_check_mark: ishakhsuvarov<br/>:white_check_mark: dmanners<br/>:white_check_mark: RomaKis<br/>:x: rossbrandon<br/>:x: olysenko<br/>:x: AlexMaxHorkun<br/>:x: aakimov<br/>:x: odubovyk<br/>:x: interio<br/>:x: dvoskoboinikov

**Comment by sidolov** [comment]
Described issue resolved in https://github.com/magento/magento2/pull/12167

**Pull request closed without merging**